server/apps/immich/src/api-v1/asset/asset.service.ts

```diff
@@ -647,9 +655,24 @@ export class AssetService {
     return this._assetRepository.getAssetCountByUserId(authUser.id);
   }
 
-  async checkAssetsAccess(authUser: AuthUserDto, assetIds: string[], mustBeOwner = false) {
+  async checkAssetsAccess(authUser: AuthUserDto, assetIds: string[], mustBeOwner = false, sharedKey?: string) {
     for (const assetId of assetIds) {
-      // Step 1: Check if user owns asset
+      // Step 1: Check if asset is part of a public shared
+      if (sharedKey) {
```

Comment:
since this operation is in a performance-critical path, I'd suggest to replace the body of this if statement with a call to a newly created custom method in AssetRepo or ShareCore that directly asks the DB to check if an asset can be accessed by a specific shared key

Reply:
Agreed, similar to how the other methods are implemented, basically returning a count and if it's 1 or more, they have access. No need to load a bunch of rows, convert to entities, etc.

Aaaaaaand, I'm pretty sure you could do this in a single query _even if_ the album assets aren't copies over. You'd just need a sub query for assets in the linked album. key with asset in (shared link asset list or shared link album asset list).